src/memory/spi_mem.c

```diff
@@ -0,0 +1,618 @@
+// Copyright 2025 Shift Crypto AG
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "spi_mem.h"
+#include "screen.h"
+#include "util.h"
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#ifndef TESTING
+#include "bitbox02_pins.h"
+#include <hal_delay.h>
+#include <spi_lite.h>
+#endif
+
+#define SECTOR_MASK 0xFFFFF000
+#define MEMORY_LIMIT (SPI_MEM_MEMORY_SIZE - 1)
+#define SR_WIP 0x01
+#define CMD_READ 0x03
+#define CMD_WREN 0x06
+#define CMD_SE 0x20
+#define CMD_PP 0x02
+#define CMD_RDSR 0x05
+#define CMD_CE 0x60
+
+static void _spi_mem_cs_low(void)
+{
+#ifndef TESTING
+    gpio_set_pin_level(PIN_MEM_CS, 0);
+#endif
+}
+
+static void _spi_mem_cs_high(void)
+{
+#ifndef TESTING
+    gpio_set_pin_level(PIN_MEM_CS, 1);
+#endif
+}
+
+static uint8_t _spi_mem_read_sr(void)
+{
+    uint8_t buffer[2] = {0};
+    buffer[0] = CMD_RDSR;
+    _spi_mem_cs_low();
+    SPI_MEM_exchange_block(buffer, 2);
```

Comment:
The calls to this function don't work in unit tests and in the simulator as you don't include the spi_lite.h header in `TESTING` (I saw a warning when compiling unit tests). This PR's CI passes because these functions are not used yet in the simulator or tests.

In my PR where I integrate everything, I could not compile the unit tests because of it.

Here is a PR to this PR where this is fixed by adding a mocked spi_mem.c to the simulator and unit tests:

https://github.com/Beerosagos/bitbox02-firmware/pull/1

Reply:
Awesome! Ty :pray:

Reply:
I was planning to merge your PR after the code here is approved, but after my last update there is a conflict. Is it ok for you to rebase once everything else looks good?